File: Competitive Programming 3/Section 1.4/Game (Card)/10205 - Stack 'em Up/Python/main.py
toSuit,toValue=["Clubs","Diamonds","Hearts","Spades"],["Jack","Queen","King","Ace"]
shuffles,cards=[],[]
def initialize(n):
    global shuffles,cards
    shuffles,cards=[],[]
    for _ in range(2):
        cards.append([])
    for _ in range(n):
        shuffles.append([])
    for i in range(52):
        cards[0].append(i)
        cards[1].append(0)
def getValue(card):
    card %= 13
    if card < 9:
        return str(card+2)
    return toValue[card-9]
def main():
    global shuffles,cards
    t = int(input())
    input()
    for cs in range(t):
        if cs > 0 :
            print()
        n = int(input())
        initialize(n)
        i,processedFromI=0,0
        while i < n:
            line = input().split()
            for tk in line:
                shuffles[i].append(int(tk)-1)
                processedFromI+=1
                if processedFromI == 52:
                    processedFromI=0
                    i+=1
        shuf = 1
        while True:
            try:
                line = input()
            except:
                break
            if line == '':
                break
            k = int(line)-1
            #shuffle
            for j in range(len(shuffles[k])):
                cards[shuf][j]=cards[1-shuf][shuffles[k][j]]
            shuf = 1 - shuf
        #! shuf has the answer
        for card in cards[1-shuf]:
            print(getValue(card)+" of "+toSuit[card//13])
main()

File: Competitive Programming 3/Section 1.4/Game (Card)/10205 - Stack 'em Up/Java/Main.java
import java.util.Scanner;
import java.util.ArrayList;

public class Main{
    static Scanner s = new Scanner(System.in);
    static ArrayList<ArrayList<Integer>> shuffles,cards; 
    static String[] toSuit = {"Clubs","Diamonds","Hearts","Spades"};
    static String[] toValue = {"Jack","Queen","King","Ace"};
    static void initialize(int n){
        shuffles = new ArrayList<ArrayList<Integer>>();
        cards = new ArrayList<ArrayList<Integer>>();
        for(int i = 0 ; i < 2 ; i ++)
            cards.add(new ArrayList<Integer>());
        for(int i = 0 ; i < n ; i ++){
            shuffles.add(new ArrayList<Integer>());
        }
        for(int i = 0; i < 52 ; i ++){
            cards.get(0).add(i);
            cards.get(1).add(0);
        }
    }
    static String getValue(int card){
        card%=13;
        if(card<9){
            return Integer.toString(card+2);
        }
        return toValue[card-9];
    }
    public static void main(String[]args){
        int t = Integer.parseInt(s.nextLine());
        s.nextLine();
        for(int cs = 0 ; cs < t; cs++){
            if(cs>0){
                System.out.printf("\n");
            }
            int n = Integer.parseInt(s.nextLine());
            initialize(n);
            for(int i = 0 , processedFromI=0;i<n;){
                String[] tks = s.nextLine().split(" ");
                for(String tk : tks){
                    shuffles.get(i).add(Integer.parseInt(tk)-1);
                    processedFromI++;
                    if(processedFromI==52){
                        processedFromI=0;
                        i++;
                    }
                }
            }
            int shuf = 1;
            while(s.hasNextLine()){
                String line = s.nextLine();
                if(line.compareTo("")==0)
                    break;
                int k = Integer.parseInt(line)-1;
                //shuffle
                int sz=shuffles.get(k).size();
                for(int j = 0 ; j < sz; j++){
                    cards.get(shuf).set(j,cards.get(1-shuf).get(shuffles.get(k).get(j))); 
                }
                shuf=1-shuf;
            }
            //! shuf has the answer
            for(int card : cards.get(1-shuf)){
                System.out.printf("%s of %s\n",getValue(card),toSuit[card/13]);
            }
        }
    }
}